components/net/tests/image_cache.rs

```diff
@@ -0,0 +1,601 @@
+        id,
+        FetchResponseMsg::ProcessResponseChunk(create_request_id(), small_chunk),
+    );
+
+    let result = cache.get_cached_image_status(url, origin, None);
+
+    match result {
+        ImageCacheResult::Pending(pending_id) => {
+            assert_eq!(id, pending_id);
+        },
+        _ => panic!("Expected Pending with insufficient data"),
+    }
+}
+
+#[test]
+fn test_notify_pending_response_with_metadata_chunk() {
+    let cache = create_test_image_cache();
+    let url = ServoUrl::parse("http://example.com/test.jpeg").unwrap();
+    let origin = mock_origin();
+
+    let id = match cache.get_cached_image_status(url.clone(), origin.clone(), None) {
+        ImageCacheResult::ReadyForRequest(id) => id,
+        _ => panic!("Expected ReadyForRequest"),
+    };
+
+    cache.notify_pending_response(
+        id,
+        FetchResponseMsg::ProcessResponse(create_request_id(), Ok(create_test_metadata())),
+    );
+
+    let jpeg_bytes = jpeg_image_bytes();
+    let metadata_chunk = jpeg_bytes[..200.min(jpeg_bytes.len())].to_vec();
+    cache.notify_pending_response(
+        id,
+        FetchResponseMsg::ProcessResponseChunk(create_request_id(), metadata_chunk),
+    );
+
+    let result = cache.get_cached_image_status(url, origin, None);
+
+    match result {
+        ImageCacheResult::Available(ImageOrMetadataAvailable::MetadataAvailable(metadata, _)) => {
+            assert!(metadata.width > 0);
+            assert!(metadata.height > 0);
+        },
+        ImageCacheResult::Pending(_) => {},
+        _ => panic!("Expected MetadataAvailable or Pending"),
+    }
+}
+
+#[test]
+fn test_notify_pending_response_complete() {
+    let cache = create_test_image_cache();
+    let url = ServoUrl::parse("http://example.com/test.jpeg").unwrap();
+    let origin = mock_origin();
+
+    let id = match cache.get_cached_image_status(url.clone(), origin.clone(), None) {
+        ImageCacheResult::ReadyForRequest(id) => id,
+        _ => panic!("Expected ReadyForRequest"),
+    };
+
+    cache.notify_pending_response(
+        id,
+        FetchResponseMsg::ProcessResponse(create_request_id(), Ok(create_test_metadata())),
+    );
+
+    let jpeg_bytes = jpeg_image_bytes();
+    cache.notify_pending_response(
+        id,
+        FetchResponseMsg::ProcessResponseChunk(create_request_id(), jpeg_bytes),
+    );
+
+    cache.notify_pending_response(
+        id,
+        FetchResponseMsg::ProcessResponseEOF(create_request_id(), Ok(create_timing())),
+    );
+
+    std::thread::sleep(std::time::Duration::from_millis(100));
+
+    let result = cache.get_cached_image_status(url.clone(), origin.clone(), None);
+
+    match result {
+        ImageCacheResult::Available(ImageOrMetadataAvailable::ImageAvailable { .. }) => {},
+        ImageCacheResult::Pending(_) => {},
+        _ => panic!("Expected ImageAvailable or Pending after complete response"),
+    }
+
+    let image = cache.get_image(url, origin, None);
+    if image.is_some() {
```

Comment:
We should be able to assert that we have an image here with my earlier suggestion.